src/stirling/source_connectors/socket_tracer/protocols/tls/parse.cc

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright 2018- The Pixie Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include "src/stirling/source_connectors/socket_tracer/protocols/tls/parse.h"
+
+#include <map>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <magic_enum.hpp>
+
+#include "src/stirling/utils/binary_decoder.h"
+
+namespace px {
+namespace stirling {
+namespace protocols {
+namespace tls {
+
+constexpr size_t kTLSRecordHeaderLength = 5;
+constexpr size_t kExtensionMinimumLength = 4;
+constexpr size_t kSNIExtensionMinimumLength = 3;
+
+// In TLS 1.3, Random is 32 bytes.
+// In TLS 1.2 and earlier, gmt_unix_time is 4 bytes and Random is 28 bytes.
+constexpr size_t kRandomStructLength = 32;
+
+StatusOr<ParseState> ExtractSNIExtension(std::map<std::string, std::string>* exts,
+                                         BinaryDecoder* decoder) {
+  PX_ASSIGN_OR(auto server_name_list_length, decoder->ExtractBEInt<uint16_t>(),
+               return ParseState::kInvalid);
+  std::vector<std::string> server_names;
+  while (server_name_list_length > 0) {
+    PX_ASSIGN_OR(auto server_name_type, decoder->ExtractBEInt<uint8_t>(),
+                 return error::Internal("Failed to extract server name type"));
+
+    // This is the only valid value for server_name_type and corresponds to host_name.
+    DCHECK_EQ(server_name_type, 0);
+
+    PX_ASSIGN_OR(auto server_name_length, decoder->ExtractBEInt<uint16_t>(),
+                 return error::Internal("Failed to extract server name length"));
+    PX_ASSIGN_OR(auto server_name, decoder->ExtractString(server_name_length),
+                 return error::Internal("Failed to extract server name"));
+
+    server_names.push_back(std::string(server_name));
+    server_name_list_length -= kSNIExtensionMinimumLength + server_name_length;
+  }
+  exts->insert({"server_name", ToJSONString(server_names)});
+  return ParseState::kSuccess;
+}
+
+ParseState ParseFullFrame(BinaryDecoder* decoder, Frame* frame) {
```

Comment:
If you have it, a pointer to the protocol's wire format specification would be useful to include here.

Reply:
Nice catch. Added in 6d17ce666.